### PR TITLE
SDIT-776: ✨ Add in summer time for business hours

### PR DESCRIPTION
**Pull request by petergphillips**
SDIT-776: ✨ Add in summer time for business hours

### CODE REVIEW

File: charts/generic-prometheus-alerts/ci/tests/rules-tests.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        eval_time: 56h
+      # 1970-01-05T09:00 UTC is Monday, working hour
+      - expr: hmpps_business_hours
+        eval_time: 105h
+        exp_samples:
+          - labels: '{__name__="hmpps_business_hours"}'
+            value: 9
+      # 1970-04-01T07:00 UTC (08:00 BST) is Monday, working hour
+      - expr: hmpps_business_hours
+        eval_time: 2167h
```
Comment (petergphillips) [inline charts/generic-prometheus-alerts/ci/tests/rules-tests.yaml:30]: this is really naff - to test the BST stuff we actually need to then wait for 2167 hours (I've upped the `evaluation_interval` to `1h`) to not make it too slow

File: charts/generic-prometheus-alerts/templates/rules.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+kind: PrometheusRule
+metadata:
+  name: {{ $targetApplication }}-rules
+  labels:
+    {{- include "generic-prometheus-alerts.labels" . | nindent 4 }}
+spec:
+  groups:
+    - name: {{ $targetApplication }}-rules
+      rules:
+        - record: summer_time_offset
```
Comment (mattops) [inline charts/generic-prometheus-alerts/templates/rules.yaml:13]: might be a good idea to namespace these record rules e.g. 
`- record: {{ $targetApplication }}:summer_time_offset`

I'm not sure how prometheus would handle duplicate recording rules with the same name.  Same for the others below.

https://prometheus.io/docs/practices/rules/#naming-and-aggregation

File: charts/generic-prometheus-alerts/templates/rules.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            or
+            (vector(1) and (month() == 10 and (day_of_month() - day_of_week()) < 25) and absent((day_of_month() >= 25) and (day_of_week() == 0)))
+            or
+            (vector(1) and ((month() == 10 and hour() < 1) or (month() == 3 and hour() > 0)) and ((day_of_month() >= 25) and (day_of_week() == 0)))
+            or
+            vector(0)
+        - record: british_time
+          expr: time() + 3600 * summer_time_offset
+        - record: hmpps_business_day
+          expr: 0 < day_of_week(british_time) < 6
```
Comment (mattops) [inline charts/generic-prometheus-alerts/templates/rules.yaml:27]: would great to make these  configurable values to the chart (with this as default), so teams can set their own working hours.  We could add this later though.

File: charts/generic-prometheus-alerts/templates/rules.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        - record: hmpps_business_hour
+          expr: 8 <= hour(british_time) < 18
```
Comment (mattops) [inline charts/generic-prometheus-alerts/templates/rules.yaml:29]: same for this, optionally set as chart value.

File: charts/generic-prometheus-alerts/templates/rules.yaml (hunk 2 above)
Comment (petergphillips) [reply charts/generic-prometheus-alerts/templates/rules.yaml:13]: that's the bit I was asking about when I asked about global (public) rules. Yes, good idea

**Review by mattops — APPROVED** [review]
LGMT

**Review by mattops — APPROVED** [review]

**Pull request merged**